apps/bc/contracts/BC.sol

```diff
@@ -0,0 +1,353 @@
+/*
+ * SPDX-License-Identitifer:    GPL-3.0-or-later
+ */
+
+pragma solidity ^0.4.24;
+
+import "@aragon/os/contracts/apps/AragonApp.sol";
+import "@aragon/os/contracts/common/EtherTokenConstant.sol";
+import "@aragon/os/contracts/common/IsContract.sol";
+import "@aragon/os/contracts/common/SafeERC20.sol";
+import "@aragon/os/contracts/lib/token/ERC20.sol";
+import "@aragon/os/contracts/lib/math/SafeMath.sol";
+
+import "@aragon/apps-token-manager/contracts/TokenManager.sol";
+import "@aragon/apps-vault/contracts/Vault.sol";
+import "@aragonblack/fundraising-pool/contracts/Pool.sol";
+
+import "./BancorContracts/converter/BancorFormula.sol";
+
+import "./IBC.sol";
+
+contract BC is EtherTokenConstant, IsContract, AragonApp, IBC, BancorFormula {
+    using SafeERC20 for ERC20;
+    using SafeMath for uint256;
+
+    uint32 public ppm = 1000000;
+    uint256 public batchBlocks;
+
+    // bytes32 public constant _ROLE = keccak256("_ROLE");
+    string private constant ERROR_POOL_NOT_CONTRACT = "BC_POOL_NOT_CONTRACT";
+    string private constant ERROR_VAULT_NOT_CONTRACT = "BC_VAULT_NOT_CONTRACT";
+    string private constant ERROR_TM_NOT_CONTRACT = "BC_TOKEN_MANAGER_NOT_CONTRACT";
+    
+    string private constant ERROR_BCTOKEN_NOT_CONTRACT = "BC_TOKEN_NOT_CONTRACT";
+    string private constant ERROR_INVALID_INIT_PARAMETER = "INVALID_INIT_PARAMETER";
+    string private constant ERROR_TRANSFERFROM_FAILED = "TRANSERFROM_FAILED";
+    string private constant ERROR_TRANSFER_FAILED = "TRANSER_FAILED";
+
+    Pool public pool;
+    Vault public vault;
+    TokenManager public tokenManager;
+
+    address public bondingCurveToken;
+
+    uint256 public MAX_COLLATERAL_TOKENS = 5;
+    address[] public collateralTokens;
+    mapping(address => uint256) public virtualSupplies;
+    mapping(address => uint256) public virtualBalances;
+    
+    mapping(address => uint256) public tokenSupplies;
```

Comment:
Shouldn't the supply be read directly from to bonded token contract? And why are their multiple token supplies? All collateral share the same bonded token, right?

Reply:
Need to discuss. Added a `poolBalance` that should maybe ask the Vault, should maybe ask the controller, should maybe be set by the controller...